File: core-java/src/main/java/com/kakao/actionbase/core/java/state/StateValue.java
package com.kakao.actionbase.core.java.state;

import com.kakao.actionbase.core.java.annotation.Nullable;
import com.kakao.actionbase.core.java.types.TypeValidator;

import org.immutables.value.Value;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

@Value.Immutable
@JsonSerialize(as = ImmutableStateValue.class)
@JsonDeserialize(as = ImmutableStateValue.class)
public interface StateValue {

  @Value.Parameter
  long version();

  @Value.Parameter
  @Nullable
  Object value();

  @Value.Check
  default void check() {
    if (value() != null) {
      TypeValidator.validateValue(value());
    }
  }

  @JsonIgnore
  default boolean isDeleted() {
    return SpecialStateValue.DELETED.code().equals(value());
  }

  @JsonIgnore
  default boolean isUnset() {
    return SpecialStateValue.UNSET.code().equals(value());
  }

  @JsonIgnore
  default boolean isPresent() {
    return value() != null && !isDeleted() && !isUnset();
  }

  static StateValue unsetOf(long version) {
    return ImmutableStateValue.builder()
        .version(version)
        .value(SpecialStateValue.UNSET.code())
        .build();
  }
}

File: core/src/main/kotlin/com/kakao/actionbase/core/state/StateValue.kt
package com.kakao.actionbase.core.state

data class StateValue(
    val version: Long,
    val value: Any?,
) {
    fun isDeleted(): Boolean = SpecialStateValue.DELETED.code() == value

    fun isUnset(): Boolean = SpecialStateValue.UNSET.code() == value

    fun isPresent(): Boolean = value != null && !isDeleted() && !isUnset()

    companion object {
        fun unsetOf(version: Long) = StateValue(version, SpecialStateValue.UNSET.code())

        fun deletedOf(version: Long) = StateValue(version, SpecialStateValue.DELETED.code())
    }
}
